pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -823,36 +1086,72 @@ func (s *DiscoveryServer) startPush(version string, push *model.PushContext, ful
 				return
 			}
 
-			select {
-			case client.pushChannel <- &XdsEvent{
+			ev := &XdsEvent{
 				push:               push,
 				pending:            &pendingPush,
 				version:            version,
 				edsUpdatedServices: edsOnly,
-			}:
-				client.LastPush = time.Now()
-				client.LastPushFailure = timeZero
-			case <-client.doneChannel: // connection was closed
-				adsLog.Infof("Client closed connection %v", client.ConID)
-			case <-time.After(PushTimeout):
-				// This may happen to some clients if the other side is in a bad state and can't receive.
-				// The tests were catching this - one of the client was not reading.
-				pushTimeouts.Add(1)
-				if client.LastPushFailure.IsZero() {
-					client.LastPushFailure = time.Now()
-					adsLog.Warnf("Failed to push, client busy %s", client.ConID)
-					pushErrors.With(prometheus.Labels{"type": "retry"}).Add(1)
-				} else {
-					if time.Since(client.LastPushFailure) > 10*time.Second {
-						adsLog.Warnf("Repeated failure to push %s", client.ConID)
-						// unfortunately grpc go doesn't allow closing (unblocking) the stream.
-						pushErrors.With(prometheus.Labels{"type": "unrecoverable"}).Add(1)
-						pushTimeoutFailures.Add(1)
-						return
+			}
+			if pilot.SingleThreadPerNode {
+				c := make(chan bool)
+				go func() {
+					// Blocking
+					s.pushConnection(client, ev)
+					c <- true
+				}()
+				select {
+				case <-c:
+					client.LastPush = time.Now()
+					client.LastPushFailure = timeZero
+				case <-client.doneChannel: // connection was closed
+					adsLog.Infof("Client closed connection %v", client.ConID)
+				case <-time.After(PushTimeout):
+					// This may happen to some clients if the other side is in a bad state and can't receive.
+					// The tests were catching this - one of the client was not reading.
+					pushTimeouts.Add(1)
+					if client.LastPushFailure.IsZero() {
+						client.LastPushFailure = time.Now()
+						adsLog.Warnf("Failed to push, client busy %s", client.ConID)
+						pushErrors.With(prometheus.Labels{"type": "retry"}).Add(1)
+					} else {
+						if time.Since(client.LastPushFailure) > 10*time.Second {
+							adsLog.Warnf("Repeated failure to push %s", client.ConID)
+							// unfortunately grpc go doesn't allow closing (unblocking) the stream.
+							pushErrors.With(prometheus.Labels{"type": "unrecoverable"}).Add(1)
+							pushTimeoutFailures.Add(1)
+							return
+						}
 					}
+
+					goto Retry
```

Comment:
I wouldn't increase PushTimeout - but I can reduce the waitAck to 1 sec.

Technically both cases should be very rare - if envoy can't process the config in 5 sec something is probably wrong. Maybe the retry logic needs to be adjusted - now we know if the push failed because of NACKs and not retry the send.